### DISCUSSION TITLE
Carthage Discontinuation

### DISCUSSION THREAD

**Discussion by paulb777** [firebase/firebase-ios-sdk#7129] [General]
We're considering discontinuing the Firebase Carthage experimental distribution in the next few months for two major reasons:

- [Carthage does not support xcframeworks](https://github.com/Carthage/Carthage/issues/2799) and as a result will not support using the Xcode simulator for binary Carthage packages on M1 Macs.
- Swift Package Manager adoption is rapidly increasing and we'd prefer to focus Firebase resources there.

Let's use this thread to discuss any concerns and migration ideas.

**Comment by mortenbekditlevsen** [comment]
Hi @paulb777 ,
I think that it is a sound choice although we are currently using Carthage at work.
I have experimented with moving to SPM, and I generally like the move.
There is however, a 'cost' related to launching an Xcode project with many dependencies like Firebase has. I assume that this will get better in future versions of Xcode, but currently it takes a while to check out the dependencies and perform resolution.
Would it be possible to include both binary targets for all Firebase frameworks as well as the full source code targets in the same Package.swift file?
At work we only update Firebase dependencies once per our release cycle, so we would be fine with binary versions. I don't know if it's possible, however.

**Reply by paulb777** [reply]
Hi @mortenbekditlevsen, 

We'll consider an SPM binary distribution in the future.  However, the SPM binary support is still immature and requires ugly workarounds.  See #6472 for details on the one Firebase component that is distributed in SPM as a binary.  

The other factor is the complexity of supporting binaries for multiple Apple platforms and Xcode versions for which the matrix is continually increasing.  Instead of providing a binary distribution, we may improve [our tooling](https://github.com/firebase/firebase-ios-sdk/tree/master/ReleaseTooling) to make it easier for developers to create their own binaries from the source distribution.

**Reply by mortenbekditlevsen** [reply]
Making it easy to roll our own binaries would be perfect! Thanks!

**Comment by lluisgerard** [comment]
I use Carthage but I already looked into moving dependencies over to SPM, I still have to find a nice way to make it work on CI environments using cached dependencies (to avoid wasting time downloading every build).
Does SPM's include `scripts/run` and `scripts/upload-symbols`? I recently came across this issue with `upload-symbols` not sending the symbols to Crashlytics on a production build. It turns out I had old scripts that were downloaded a long time ago, that failed to upload the desymbolication. All I had to do is update the binary, which is something we forgot because there was no update mechanism from our side.
If this makes that problem disappear, so you can have the binaries included, it's a big win for us.

**Reply by paulb777** [reply]
Hi @lluisgerard 

We haven't found a good way to install the Crashlytics scripts via SPM. We are investigating options for this and there is an open issue at #6864.

**Comment by ghost** [comment]
@paulb777 https://github.com/Carthage/Carthage/pull/3071

**Reply by paulb777** [reply]
Thanks for the link @fr-machlin . It's not clear to me that this addresses using binary xcframeworks versus creating them. See https://github.com/Carthage/Carthage/pull/3071#issuecomment-733180109.

**Reply by paulb777** [reply]
More at https://github.com/Carthage/Carthage/pull/3071#issuecomment-747085975

**Reply by CraigSiemens** [reply]
That PR has been merged in and they've mentioned in those comments [and others](https://github.com/Carthage/Carthage/issues/3102#issuecomment-755752392) that binary support is coming.

**Comment by jafara** [comment]
yes please get rid of it. carthage is outdated

**Comment by schlingding** [comment]
Other enterprise level frameworks still supporting Carthage are continuing to have issues supporting SPM (MSAL and DJI-SDK are a few examples) which would suggest moving the timeline on discontinuation until SPM matures further.

**Reply by paulb777** [reply]
Agreed, we'll keep it going along as the value continues and the cost to maintain doesn't get too expensive. The lack for Carthage support for binary xcframeworks and ARM-based Macs might drive the discontinuation sooner than the growth of SPM.

**Reply by RajaveluC** [reply]
Hi @paulb777 , The latest version of Carthage has support for XCFramework. 
Ref: https://github.com/Carthage/Carthage/releases/tag/0.37.0

**Reply by paulb777** [reply]
Hi @RajaveluC See the [discussion above](https://github.com/firebase/firebase-ios-sdk/discussions/7129#discussioncomment-203613). Carthage is still missing support for binary `xcframework` distribution.

**Reply by pbush25** [reply]
How much work will it take to make Firebase available with this version of Carthage once they release support for binary xcframeworks which they say is coming? 

"Support for binary xcframeworks will be added in a future release. For now, continue integrating binary dependencies using the strategy of platform-specific frameworks."

Moving to SPM for one dependency isn't really ideal for us at the moment, but I'd be happy to try and contribute to whatever is needed to get the XCFrameworks binaries working when Carthage releases support.

**Reply by paulb777** [reply]
Another workaround is to use the [Firebase zip distribution](https://github.com/firebase/firebase-ios-sdk/releases/tag/7.6.0) which is now comprised of `xcframeworks`.

It will be relatively straight forward to update https://github.com/firebase/firebase-ios-sdk/tree/master/ReleaseTooling to build a Carthage distribution of the same `xcframeworks` once Carthage supports it.

**Reply by jesusmateos1234** [reply]
Carthage supports XCFrameworks

**Reply by sebskuse** [reply]
Adding to this, there is an open PR to add binary support to Carthage: https://github.com/Carthage/Carthage/pull/3123.

I think we're in a similar situation to a lot of other Carthage users here: We'd love to move to something like SPM, but the penalty of having to rebuild all of our dependencies for every integration on CI is just too much.

**Comment by agirault** [comment]
Is there a timeline for adding `Performance` to SPM? I see it listed as a [limitation](https://github.com/firebase/firebase-ios-sdk/blob/master/SwiftPackageManager.md#limitations). This is the only reason we still use Carthage.

**Reply by paulb777** [reply]
We're working on it, but don't publish future timelines. It's tracked at #6528

**Comment by Lavmint** [comment]
I came here in search of how to get rid of firebase rebuilding on every build of a project. 
I found out that Carthage has xcframework option and firebase now is also provided as xcframeworks.

I wanted to add xcframework as a remote binary dependency to SPM. 
However it seems that firebase provides xcframeworks packed in zip (so I can't address them directly in Package.swift) and also dropping support of Carthage (that successfully built KeychainSwift for me)

It feels like firebase trying to bully me and wants to be builded again and again.
Is there any other options available?

**Reply by paulb777** [reply]
@Lavmint [Carthage binary xcframework support just recently merged](https://github.com/Carthage/Carthage/pull/3123) and has not yet released yet.

See the related issue about Swift Package Manager binary distribution at #6564

In short, there are two current workarounds:
- Reorganize the xcframeworks from the zip 
- Rebuild the binaries however you like with this [zip-builder](https://github.com/firebase/firebase-ios-sdk/tree/master/ReleaseTooling#zip-builder)

**Reply by agirault** [reply]
It ~~'s~~ should be released as part of 0.38.0: https://github.com/Carthage/Carthage/pull/3152

**Comment by agirault** [comment]
@paulb777 Has support for Carthage been dropped already? I tried to fetch Firebase 7.10 with Carthage but the resolved version numbers I get are mismatched for some modules:

```
binary "https://dl.google.com/dl/firebase/ios/carthage/FirebaseAnalyticsBinary.json" "7.4.0"
binary "https://dl.google.com/dl/firebase/ios/carthage/FirebaseCrashlyticsBinary.json" "7.10.0"
binary "https://dl.google.com/dl/firebase/ios/carthage/FirebasePerformanceBinary.json" "7.10.0"
binary "https://dl.google.com/dl/firebase/ios/carthage/FirebaseProtobufBinary.json" "7.6.0"
```

I also don't see Carthage mentioned here: https://firebase.google.com/docs/ios/setup#integrate-without-cocoapods

**Reply by paulb777** [reply]
Carthage support has always been experimental and not part of the official product docs. See https://github.com/firebase/firebase-ios-sdk/blob/master/Carthage.md. Some of the library updates are blocked on the Carthage 0.38.0 release. Details in #7039.  There are not near-term plans to discontinue the experimental support.

**Reply by agirault** [reply]
Thanks. Just found the explanation here as well: https://github.com/firebase/firebase-ios-sdk/issues/7698

**Reply by Lavmint** [reply]
Why do you have different versions for each binary if firebase expects the same version for all?
I successfully integrated firebase with Carthage again on another project and I'm happy with it till .xcframeworks would be available with SPM. However we are still using version 6.33

**Reply by paulb777** [reply]
While it is not always true that different Firebase versions are compatible with each other, the latest Carthage versions are.

**Reply by agirault** [reply]
Could you confirm what are the latest Firebase versions we can safely use with Carthage 0.38.0 if they differ across modules?

**Reply by paulb777** [reply]
See #7946 for some detail about versioning and follow #7039 to track progress for restoring the Carthage distribution with the latest Firebase distributions

**Comment by paulb777** [comment]
Now that Carthage 0.38.0 is supporting binary xcframeworks, we've restored the experimental Carthage distribution with the 8.0.0 release. 

We'll keep it going for the foreseeable future.

**Reply by agirault** [reply]
Fantastic, thank you!

I have a couple of questions/remarks, let me know if you'd rather I open a separate issue (this discussion/megathread seems handy for this as well):

Command to reproduce: `carthage bootstrap --platform ios --use-xcframeworks`

1. Is Protobuf still at `7.6.0` (no `8.0.0` release)? 
    <img width="1072" alt="Screen Shot 2021-05-20 at 11 05 23 AM" src="https://user-images.githubusercontent.com/4910855/119002919-496d6280-b95b-11eb-9841-d96d577c589c.png">
2. `Firebase` is still a `.framework`. `Protobuf` is built both as a `.framework` and a `.xcframework`:
    <img width="496" alt="Screen Shot 2021-05-20 at 11 11 35 AM" src="https://user-images.githubusercontent.com/4910855/119003964-255e5100-b95c-11eb-8b5b-ea8861fcdb8e.png">
3. The `xcframework`s include `ios`, `macos` and `tvos` frameworks, even when passing ` --platform ios` to `carthage bootstrap`:
    <img width="589" alt="Screen Shot 2021-05-20 at 11 04 01 AM" src="https://user-images.githubusercontent.com/4910855/119002686-16c36a00-b95b-11eb-89d9-c5019cced883.png">

**Reply by paulb777** [reply]
@agirault Thanks for taking a look and providing the feedback and questions.

1 & 2. FirebaseProtobufBinary.json should no longer be used. Firebase is migrating from using the Protobuf library and it's currently only left in FirebasePerformance from which its `.xcframework` can be found.

3. Most of the `.xcframeworks` include `macos` and `tvos` support. We're currently not documenting it since it is untested and missing  the Analytics xcframeworks and the Firebase framework.  Having them stripped on installation with the `--platform ios` option is probably best addressed as a feature request to the Carthage project.

**Reply by haithngn** [reply]
@paulb777 in Xcode 12.5 I cannot build for the real device or for archiving because Firebase.framework file not found for architecture arm64, I already tried:
- Validate Workspace: YES ~> does not work.
- Exclude architectures: arm64 -> does not work.
- Remove Firebase.framework then manual import frameworks inside Bridge Header file ~~> does not work.

Can we have a Firebase.xcframework? any workaround?

Tks,

**Reply by paulb777** [reply]
@haithngn I'm not able to reproduce when following the instructions at https://github.com/firebase/firebase-ios-sdk/blob/master/Carthage.md. Are you deleting Firebase.framework from the Link Binary with Libraries Build Phase - https://github.com/firebase/firebase-ios-sdk/issues/911#issuecomment-372455235

**Reply by haithngn** [reply]
@paulb777 thank you for a quick reply,

Create new project in the Xcode version 12.5, Carthage version 0.3.8,
![Screen Shot 2021-06-13 at 1 16 39 PM](https://user-images.githubusercontent.com/22902671/121797272-70ad0c00-cc49-11eb-94fd-2770ba032800.png)


![Screen Shot 2021-06-13 at 1 17 35 PM](https://user-images.githubusercontent.com/22902671/121797292-94705200-cc49-11eb-8632-0a4299604e4b.png)

(https://github.com/firebase/firebase-ios-sdk/issues/911#issuecomment-372455235) is outdated, he used older Carthage which generates the .framework file

**Reply by paulb777** [reply]
FIRApp comes from FirebaseCore.xcframework which should be in the `Link Binary With Libraries` Build Phase.
